citylots_bench_test.go

```diff
@@ -0,0 +1,55 @@
+package quamina
+
+import (
+	"testing"
+)
+
+func BenchmarkCityLots(b *testing.B) {
```

Comment:
IMHO the benchmark looks good to avoid compiler optimizations, but just to double check did you run with `go test -gcflags "-m -m" -bench=<...>` and check for inlining?

Reply:
Didn't double checked, but now I checked and I don't see inlining here -

```
./citylots_bench_test.go:48:42: err.Error() escapes to heap:
./citylots_bench_test.go:48:42:   flow: {storage for ... argument} = &{storage for err.Error()}:
./citylots_bench_test.go:48:42:     from err.Error() (spill) at ./citylots_bench_test.go:48:42
./citylots_bench_test.go:48:42:     from ... argument (slice-literal-element) at ./citylots_bench_test.go:48:12
./citylots_bench_test.go:48:42:   flow: {heap} = {storage for ... argument}:
./citylots_bench_test.go:48:42:     from ... argument (spill) at ./citylots_bench_test.go:48:12
./citylots_bench_test.go:48:42:     from (*testing.common).Errorf(b.common, "Matches4JSON: %s", ... argument...) (call parameter) at ./citylots_bench_test.go:48:12
./citylots_bench_test.go:34:20: matcherStats(q.matcher.(*coreMatcher)) escapes to heap:
./citylots_bench_test.go:34:20:   flow: {storage for ... argument} = &{storage for matcherStats(q.matcher.(*coreMatcher))}:
./citylots_bench_test.go:34:20:     from matcherStats(q.matcher.(*coreMatcher)) (spill) at ./citylots_bench_test.go:34:20
./citylots_bench_test.go:34:20:     from ... argument (slice-literal-element) at ./citylots_bench_test.go:34:7
./citylots_bench_test.go:34:20:   flow: {heap} = {storage for ... argument}:
./citylots_bench_test.go:34:20:     from ... argument (spill) at ./citylots_bench_test.go:34:7
./citylots_bench_test.go:34:20:     from (*testing.common).Log(b.common, ... argument...) (call parameter) at ./citylots_bench_test.go:34:7
./citylots_bench_test.go:31:47: err.Error() escapes to heap:
./citylots_bench_test.go:31:47:   flow: {storage for ... argument} = &{storage for err.Error()}:
./citylots_bench_test.go:31:47:     from err.Error() (spill) at ./citylots_bench_test.go:31:47
./citylots_bench_test.go:31:47:     from ... argument (slice-literal-element) at ./citylots_bench_test.go:31:12
./citylots_bench_test.go:31:47:   flow: {heap} = {storage for ... argument}:
./citylots_bench_test.go:31:47:     from ... argument (spill) at ./citylots_bench_test.go:31:12
./citylots_bench_test.go:31:47:     from (*testing.common).Fatalf(b.common, "AddPattern failed: %s", ... argument...) (call parameter) at ./citylots_bench_test.go:31:12
./citylots_bench_test.go:29:27: names[i] escapes to heap:
./citylots_bench_test.go:29:27:   flow: x = &{storage for names[i]}:
./citylots_bench_test.go:29:27:     from names[i] (spill) at ./citylots_bench_test.go:29:27
./citylots_bench_test.go:29:27:     from q, x, patternJSON := q, names[i], patterns[i] (assign-pair) at ./citylots_bench_test.go:29:21
./citylots_bench_test.go:29:27:   flow: {heap} = x:
./citylots_bench_test.go:29:27:     from q.matcher.addPattern(x, patternJSON) (call parameter) at ./citylots_bench_test.go:29:21
./citylots_bench_test.go:26:34: err.Error() escapes to heap:
./citylots_bench_test.go:26:34:   flow: {storage for ... argument} = &{storage for err.Error()}:
./citylots_bench_test.go:26:34:     from err.Error() (spill) at ./citylots_bench_test.go:26:34
./citylots_bench_test.go:26:34:     from ... argument (slice-literal-element) at ./citylots_bench_test.go:26:11
./citylots_bench_test.go:26:34:   flow: {heap} = {storage for ... argument}:
./citylots_bench_test.go:26:34:     from ... argument (spill) at ./citylots_bench_test.go:26:11
./citylots_bench_test.go:26:34:     from (*testing.common).Fatalf(b.common, "New(): %s", ... argument...) (call parameter) at ./citylots_bench_test.go:26:11
./citylots_bench_test.go:7:24: parameter b leaks to {heap} with derefs=0:
./citylots_bench_test.go:7:24:   flow: {heap} = b:
./citylots_bench_test.go:7:24:     from b.common (dot of pointer) at ./citylots_bench_test.go:26:4
./citylots_bench_test.go:7:24:     from b.common (address-of) at ./citylots_bench_test.go:26:4
./citylots_bench_test.go:7:24:     from (*testing.common).Fatalf(b.common, "New(): %s", ... argument...) (call parameter) at ./citylots_bench_test.go:26:11
./citylots_bench_test.go:7:24: leaking param: b
./citylots_bench_test.go:10:22: []string{...} does not escape
./citylots_bench_test.go:16:19: []string{...} does not escape
./citylots_bench_test.go:26:11: ... argument does not escape
./citylots_bench_test.go:26:34: err.Error() escapes to heap
./citylots_bench_test.go:29:27: names[i] escapes to heap
./citylots_bench_test.go:31:12: ... argument does not escape
./citylots_bench_test.go:31:47: err.Error() escapes to heap
./citylots_bench_test.go:34:7: ... argument does not escape
./citylots_bench_test.go:34:20: matcherStats(q.matcher.(*coreMatcher)) escapes to heap
./citylots_bench_test.go:48:12: ... argument does not escape
./citylots_bench_test.go:48:42: err.Error() escapes to heap
```

Reply:
Why would we want to avoid compiler optimizations?  The benchmark should give a realistic read-out of what performance people should expect?

Reply:
@timbray you want to avoid compiler optimisations in the benchmark it self, so the compiler for example won't eliminate the benchmark loop, for example:
```
func BenchmarkFib20(b *testing.B) {
	for i := 0; i < b.N; i++ {
		Fib(20)
	}
}
```

In this case, the compiler can eliminate the calls to "Fib(20)", so you want try eliminate those issues (but not in the code under benchmark, in our case the matcher/flattener)

Reply:
Exactly @yosiat ! Thx for confirming.